engine/chunker/chain_chunker.go

```diff
@@ -0,0 +1,101 @@
+package chunker
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	provider "github.com/filecoin-project/index-provider"
+	"github.com/filecoin-project/storetheindex/api/v0/ingest/schema"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/multiformats/go-multihash"
+)
+
+var _ EntriesChunker = (*ChainChunker)(nil)
+
+// ChainChunker chunks advertisement entries as a chained series of schema.EntryChunk nodes.
+// See: NewChainChunker
+type ChainChunker struct {
+	ls        *ipld.LinkSystem
+	chunkSize int
+}
+
+// NewChainChunker instantiates a new HAMT chunker that given a provider.MultihashIterator it drains
+// all its mulithashes and stores them in the given link system represented as a chain of
+// schema.EntryChunk nodes where each chunk contains no more than chunkSize number of multihashes.
+//
+// See: schema.EntryChunk.
+func NewChainChunker(ls *ipld.LinkSystem, chunkSize int) (*ChainChunker, error) {
+	if chunkSize < 1 {
+		return nil, fmt.Errorf("chunk size must be at least 1; got: %d", chunkSize)
+	}
+	return &ChainChunker{
+		ls:        ls,
+		chunkSize: chunkSize,
+	}, nil
+}
+
+func NewChainChunkerFunc(chunkSize int) NewChunkerFunc {
+	return func(ls *ipld.LinkSystem) (EntriesChunker, error) {
+		return NewChainChunker(ls, chunkSize)
+	}
+}
+
+// Chunk chunks all the mulithashes returned by the given iterator into a chain of schema.EntryChunk
+// nodes where each chunk contains no more than chunkSize number of multihashes and returns the link
+// the root chunk node.
+//
+// See: schema.EntryChunk.
+func (ls *ChainChunker) Chunk(ctx context.Context, mhi provider.MultihashIterator) (ipld.Link, error) {
+	mhs := make([]multihash.Multihash, 0, ls.chunkSize)
+	var next ipld.Link
+	var mhCount, chunkCount int
+	for {
+		mh, err := mhi.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
```

Comment:
I'm curious do you find this style clearer to read? The rationale for writing it the way I did is to have one less extra check for EOF case since `nil` will not equal `EOF`